### PR TITLE
Fix unreachable validation for gradcheck

### PR DESCRIPTION
**Pull request by takp**
Hi, I found the validation that is unreachable in `gradcheck` function :)

### CODE REVIEW

**Comment by albanD** [comment]
Hi,

Thanks for catching this.
Your change seems to be braking many tests though.

**Comment by takp** [comment]
That's true... I'll try to update the tests.

**Comment by takp** [comment]
@albanD Would you please give me an advise? 

- One of the test affected: https://github.com/pytorch/pytorch/blob/master/test/test_autograd.py#L5654

This `test_where_functional` executes `gradcheck`, but the `gradcheck` requires the all input should have `requires_grad=True`.  On the other hand, `cond` (one of the input) can't have `requires_grad=True` because `cond` is boolean matrix, not a floating point dtype. 

I think we should choose one of the following:

- Stop testing  `gradcheck` for the tests that are not incompatible with its validation of `requires_grad=True`
- Stop raising exception for this validation? (if `some_input_not_requiring_grad`)

**Comment by albanD** [comment]
The current gradcheck code seems to handle the case where some inputs do not require gradients.
So just removing the check sounds ok to me.

**Review by albanD — APPROVED** [review]
Thanks !

The complex errors are most likely because you rebased on master when it was broken this morning :/ 
could you rebase again please to make sure all is good?

**Comment by takp** [comment]
I merged the current master branch, but it fails with 1 test... 🤔

**Comment by albanD** [comment]
No worries, this one is a known issue ;)
